apps/prairielearn/src/middlewares/authzCourseOrInstance.ts

```diff
@@ -205,420 +471,276 @@ export async function authzCourseOrInstance(req: Request, res: Response) {
-    debug('effective user was denied access');
-
-    res.locals.authz_data.user = user;
-    res.locals.authz_data.is_administrator = false;
+    // If we didn't throw an error and we can't set the effective user, return all nulls.
+    if (verifyAuthnAuthResultResult.state === 'ignore_override') {
+      // If we can't set the effective user, the effective user is the same as the authn user.
+      return {
+        authResult: authnAuthResult,
+        course: authnCourse,
+        institution: authnInstitution,
+        courseInstance: authnCourseInstance,
+      };
+    }
+    assert(verifyAuthnAuthResultResult.state === 'allow_override');
+
+    // We are trying to override the user data.
+    debug('trying to override the user data');
+    debug(req.cookies);
+
+    const {
+      authResult: effectiveAuthResult,
+      course: effectiveCourse,
+      institution: effectiveInstitution,
+      courseInstance: effectiveCourseInstance,
+    } = await calculateAuthData({
+      user: effectiveUserData ? effectiveUserData.user : authnAuthResult.user,
+      course_id: req.params.course_id || null,
+      course_instance_id: req.params.course_instance_id || null,
+      ip: req.ip || null,
+      req_date,
+      overrides: {
+        is_administrator: effectiveUserData
+          ? effectiveUserData.is_administrator
+          : res.locals.is_administrator,
+        allow_example_course_override: false,
+        req_mode: config.devMode ? req.cookies.pl_test_mode : null,
+        req_course_role: req.cookies.pl2_requested_course_role || null,
+        req_course_instance_role: req.cookies.pl2_requested_course_instance_role || null,
+      },
+    });
 
-    res.locals.authz_data.course_role = 'None';
-    res.locals.authz_data.has_course_permission_preview = false;
-    res.locals.authz_data.has_course_permission_view = false;
-    res.locals.authz_data.has_course_permission_edit = false;
-    res.locals.authz_data.has_course_permission_own = false;
+    // If the authn user were denied access, then we would return an error. Here,
+    // we simply return (without error). This allows the authn user to keep access
+    // to pages (e.g., the effective user page) for which only authn permissions
+    // are required.
+    if (effectiveAuthResult === null) {
+      return {
+        authResult: {
+          user: effectiveUserData ? effectiveUserData.user : authnAuthResult.user,
+          is_administrator: false,
+          course_role: 'None' as EnumCourseRole,
+          mode: authnAuthResult.mode,
+          mode_reason: authnAuthResult.mode_reason,
+          ...calculateCourseRolePermissions('None'),
+          ...(req.params.course_instance_id
+            ? {
+                course_instance_role: 'None' as EnumCourseInstanceRole,
+                has_student_access: false,
+                has_student_access_with_enrollment: false,
+                ...calculateCourseInstanceRolePermissions('None'),
+              }
+            : {}),
```

Comment:
Nit, keeping course-role-related properties next to each other:

```suggestion
          course_role: 'None' as EnumCourseRole,
          ...calculateCourseRolePermissions('None'),
          ...(req.params.course_instance_id
            ? {
                course_instance_role: 'None' as EnumCourseInstanceRole,
                has_student_access: false,
                has_student_access_with_enrollment: false,
                ...calculateCourseInstanceRolePermissions('None'),
              }
            : {}),
          mode: authnAuthResult.mode,
          mode_reason: authnAuthResult.mode_reason,
```